openbas-api/src/main/java/io/openbas/service/ScenarioService.java

```diff
@@ -119,6 +130,78 @@ public Page<Scenario> scenarios(SearchPaginationInput searchPaginationInput) {
+      Pageable pageable) {
+    CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+
+    // -- Create Query --
+    CriteriaQuery<Tuple> cq = cb.createTupleQuery();
+    // FROM
+    Root<Scenario> scenarioRoot = cq.from(Scenario.class);
+    // Join on TAG
+    Join<Scenario, Tag> scenarioTagsJoin = scenarioRoot.join("tags", JoinType.LEFT);
+    Expression<String[]> tagIdsExpression = cb.function("array_agg", String[].class, scenarioTagsJoin.get("id"));
+    // Join on INJECT and INJECTOR CONTRACT
+    Join<Scenario, Inject> injectsJoin = scenarioRoot.join("injects", JoinType.LEFT);
+    Join<Inject, InjectorContract> injectorsContractsJoin = injectsJoin.join("injectorContract", JoinType.LEFT);
+    // SELECT
+    cq.multiselect(
+        scenarioRoot.get("id").alias("scenario_id"),
+        scenarioRoot.get("name").alias("scenario_name"),
+        scenarioRoot.get("severity").alias("scenario_severity"),
+        scenarioRoot.get("category").alias("scenario_category"),
+        scenarioRoot.get("recurrence").alias("scenario_recurrence"),
+        scenarioRoot.get("updatedAt").alias("scenario_updated_at"),
+        tagIdsExpression.alias("scenario_tags"),
+        injectorsContractsJoin.get("platforms").alias("scenario_platforms")
+    ).distinct(true);
+    // Group By
+    cq.groupBy(scenarioRoot.get("id"), injectorsContractsJoin.get("platforms"));
+
+    // -- Text Search and Filters --
+    if (specification != null) {
+      Predicate predicate = specification.toPredicate(scenarioRoot, cq, cb);
+      if (predicate != null) {
+        cq.where(predicate);
+      }
+    }
+
+    // -- Sorting --
+    List<Order> orders = toSortCriteriaBuilder(cb, scenarioRoot, pageable.getSort());
+    cq.orderBy(orders);
+
+    // Type Query
+    TypedQuery<Tuple> query = entityManager.createQuery(cq);
+
+    // -- Pagination --
+    query.setFirstResult((int) pageable.getOffset());
+    query.setMaxResults(pageable.getPageSize());
+
+    // -- EXECUTION --
+    List<RawPaginationScenario> scenarios = query.getResultList()
+        .stream()
+        .map(tuple -> new RawPaginationScenario(
+            tuple.get("scenario_id", String.class),
+            tuple.get("scenario_name", String.class),
+            tuple.get("scenario_severity", String.class),
+            tuple.get("scenario_category", String.class),
+            tuple.get("scenario_recurrence", String.class),
+            tuple.get("scenario_updated_at", Instant.class),
+            tuple.get("scenario_tags", String[].class),
+            tuple.get("scenario_platforms", String[].class)
+        ))
+        .toList();
+
+    // -- Count Query --
+    CriteriaQuery<Long> countQuery = cb.createQuery(Long.class);
+    Root<Scenario> countRoot = countQuery.from(Scenario.class);
+    countQuery.select(cb.count(countRoot));
+    Long total = entityManager.createQuery(countQuery).getSingleResult();
+
+    return new PageImpl<>(scenarios, pageable, total);
```

Comment:
Couldn't we skip the second query and just do a dumb .size() on the scenarios ?

Reply:
scenarios contains only the X first element depending of the pagination